python/dgl/nn/mxnet/conv/gatconv.py

```diff
@@ -84,33 +187,69 @@ def __init__(self,
+        Set allow_zero_in_degree flag.
+
+        Parameters
+        ----------
+        set_value : bool
+            The value to be set to the flag.
+        """
+        self._allow_zero_in_degree = set_value
+
     def forward(self, graph, feat):
-        r"""Compute graph attention network layer.
+        r"""
+
+        Description
+        -----------
+        Compute graph attention network layer.
 
         Parameters
         ----------
         graph : DGLGraph
             The graph.
-        feat : mxnet.NDArray
-            If a mxnet.NDArray is given, the input feature of shape :math:`(N, D_{in})`
-            where :math:`D_{in}` is size of input feature, :math:`N` is the number of
-            nodes.
-            If a pair of mxnet.NDArray is given, the pair must contain two tensors of
-            shape :math:`(N_{in}, D_{in_{src}})` and :math:`(N_{out}, D_{in_{dst}})`.
+        feat : mxnet.NDArray or pair of mxnet.NDArray
+            If a mxnet.NDArray is given, the input feature of shape :math:`(N, D_{in})` where
+            :math:`D_{in}` is size of input feature, :math:`N` is the number of nodes.
+            If a pair of mxnet.NDArray is given, the pair must contain two tensors of shape
+            :math:`(N_{in}, D_{in_{src}})` and :math:`(N_{out}, D_{in_{dst}})`.
 
         Returns
         -------
         mxnet.NDArray
             The output feature of shape :math:`(N, H, D_{out})` where :math:`H`
             is the number of heads, and :math:`D_{out}` is size of output feature.
+
+        Raises
+        ------
+        DGLError
+            If there are 0-in-degree nodes in the input graph, it will raise DGLError
+            since no message will be passed to those nodes. This will cause invalid output.
+            The error can be ignored by setting ``allow_zero_in_degree`` parameter to ``True``.
         """
         with graph.local_scope():
+            if not self._allow_zero_in_degree:
+                if graph.in_degrees().min() == 0:
+                    raise DGLError('There are 0-in-degree nodes in the graph, '
+                                   'output for those nodes will be invalid. '
+                                   'This is harmful for some applications, '
+                                   'causing silent performance regression. '
+                                   'Adding self-loop on the input graph by '
+                                   'calling `g = dgl.add_self_loop(g)` will resolve '
+                                   'the issue. Setting ``allow_zero_in_degree`` '
+                                   'to be `True` when constructing this module will '
+                                   'suppress the check and let the code run.')
+
             if isinstance(feat, tuple):
                 h_src = self.feat_drop(feat[0])
                 h_dst = self.feat_drop(feat[1])
-                feat_src = self.fc(h_src).reshape(
+                feat_src = self.fc_src(h_src).reshape(
```

Comment:
Considering there are cases we apply GAT-on-homograph to bipartite graph, I suggest add a if-statement here:
```suggestion
                if not has_attr(self, fc_src): self.fc_src, self.fc_dst = self.fc, self.fc
                feat_src = self.fc_src(h_src).reshape(
```

Reply:
When will we apply a GAT-on-homo to bipartite graph? As discussed before, for sample-based inference, we will just is_block() and  graph.number_of_dst_nodes() to get the dst nodes, no need to make the input tuple.